server/src/internalClusterTest/java/org/elasticsearch/indices/recovery/SnapshotBasedIndexRecoveryIT.java

```diff
@@ -0,0 +1,752 @@
+        String indexName = randomAlphaOfLength(10).toLowerCase(Locale.ROOT);
+        createIndex(indexName,
+            Settings.builder()
+                .put(IndexMetadata.SETTING_NUMBER_OF_SHARDS, 1)
+                .put(IndexMetadata.SETTING_NUMBER_OF_REPLICAS, 0)
+                .put(MergePolicyConfig.INDEX_MERGE_ENABLED, false)
+                .put(IndexService.GLOBAL_CHECKPOINT_SYNC_INTERVAL_SETTING.getKey(), "1s")
+                .put("index.routing.allocation.require._name", sourceNode)
+                .build()
+        );
+
+        int numDocs = randomIntBetween(3000, 5000);
+        indexDocs(indexName, 0, numDocs);
+        forceMerge();
+
+        String repoName = "repo";
+        createRepo(repoName, TestRepositoryPlugin.INSTRUMENTED_TYPE);
+        createSnapshot(repoName, "snap", Collections.singletonList(indexName));
+
+        int docsIndexedAfterSnapshot = randomIntBetween(1000, 2000);
+        indexDocs(indexName, numDocs, docsIndexedAfterSnapshot);
+
+        String targetNode = internalCluster().startDataOnlyNode();
+        assertAcked(
+            client().admin().indices().prepareUpdateSettings(indexName)
+                .setSettings(Settings.builder()
+                    .put("index.routing.allocation.require._name", targetNode)).get()
+        );
+
+        ensureGreen();
+
+        RecoveryState recoveryState = getLatestPeerRecoveryStateForShard(indexName, 0);
+        assertPeerRecoveryWasSuccessful(recoveryState, sourceNode, targetNode);
+
+        InstrumentedRepo repository = getRepositoryOnNode(repoName, targetNode);
+
+        long snapshotSizeForIndex = getSnapshotSizeForIndex(repoName, "snap", indexName);
+        assertThat(repository.totalBytesRead.get(), is(greaterThan(0L)));
+        assertThat(repository.totalBytesRead.get(), is(lessThanOrEqualTo(snapshotSizeForIndex)));
+
+        assertDocumentsAreEqual(indexName, numDocs + docsIndexedAfterSnapshot);
+    }
+
+    public void testPeerRecoveryDoNotUseSnapshotsWhenSegmentsAreNotShared() throws Exception {
+        String sourceNode = internalCluster().startDataOnlyNode();
+        String indexName = randomAlphaOfLength(10).toLowerCase(Locale.ROOT);
+        createIndex(indexName,
+            Settings.builder()
+                .put(IndexMetadata.SETTING_NUMBER_OF_SHARDS, 1)
+                .put(IndexMetadata.SETTING_NUMBER_OF_REPLICAS, 0)
+                .put(IndexService.GLOBAL_CHECKPOINT_SYNC_INTERVAL_SETTING.getKey(), "1s")
+                .put("index.routing.allocation.require._name", sourceNode)
+                .build()
+        );
+
+        int numDocs = randomIntBetween(3000, 5000);
+        indexDocs(indexName, 0, numDocs);
+
+        String repoName = "repo";
+        createRepo(repoName, TestRepositoryPlugin.INSTRUMENTED_TYPE);
+        createSnapshot(repoName, "snap", Collections.singletonList(indexName));
+
+        int docsIndexedAfterSnapshot = randomIntBetween(1000, 2000);
```

Comment:
Can we lower the lower bound perhaps to 1?

```suggestion
        int docsIndexedAfterSnapshot = randomIntBetween(1, 2000);
```